patroni/postgresql.py

```diff
@@ -733,6 +738,36 @@ def get_pid(self):
             logger.warning("Garbage pid in postmaster.pid: {0!r}".format(pid))
             return 0
 
+    def get_pid_with_lost_data_dir(self):
+        process = list(filter(
+            lambda p: p.pid == self.postmaster["pid"] and p.name() == "postgres" and self._data_dir in p.cmdline(),
```

Comment:
We already know the exact pid, no need to iterate through all processes in the system.

Just replace this line with:
```python
process = psutil.Process(self._postmaster_info['pid'])
```

`self._data_dir in p.cmdline()` -- it will check only exact match, and unfortunately can't handle all cases:
 postgresql.data_dir contains absolute path (/home/postgres/data), but postgres was started with `-D data` arguments and other way around.
I would suggest to use `process.cwd()` instead. This value always match with the real data directory.
Be careful, `cwd` method can throw `psutil.AccessDenied` exception. Also when comparing `process.cwd()` with `self._data_dir`, don't forget that you need to compare absolute paths.

Reply:
I think it would be best to cache the psutil.Process object once we know it's the correct process (`create_time()` matches start_time from pid file). psutil.Process methods guarantee that we have the correct process for [the methods that interest us](https://github.com/giampaolo/psutil/blob/62c845a7d97eea0c2740b8f488a62b6c08eb4331/psutil/__init__.py#L378-L397). This requires a bit more refactoring, but I suspect that it will give a net win in simplicity.

Reply:
We can have another process with same id if previos was killed.  `p.cmdline()` returns array with parameters of command like `['/usr/pgsql-9.6/bin/postgres', '-D', '/var/lib/pgsql/data/postgresql_node1', .... ]` so it will be safe search for `self._data_dir` which should be passed to postgresql, if I understand correctly.

Reply:
Also I cannot guarantee if process still running so it's more safe to check if process exists than call `psutil.Process()`

Reply:
@ants @CyberDem0n should we leave current version or change to `psutil.Process(self._postmaster_info['pid'])` ?